File: android/core/common/src/main/kotlin/com/fatwabot/core/common/DeepLink.kt
package com.fatwabot.core.common

import android.net.Uri

/**
 * Canonical in-app destinations reachable from outside the app — today from
 * widgets, tomorrow from push payloads or a website.
 *
 * Lives in `:core:common` so the `:app` module and the `:widget` module build
 * their URIs from the SAME definition. A widget that hardcodes
 * `"fatwabot://prayer"` while the app parses `"prayer-times"` fails silently —
 * the tap just opens Home and nobody notices for weeks — so the scheme and its
 * parsing are defined exactly once, here.
 *
 * Mirrors iOS `CoreKit.DeepLink`; keep the two in sync, and keep [SCHEME] in
 * sync with the intent-filter in the app's AndroidManifest.
 */
enum class DeepLink(val host: String) {
    HOME("home"),
    PRAYER("prayer"),
    QIBLA("qibla"),
    TASBEEH("tasbeeh"),
    AZKAR("azkar"),
    DUA("dua"),
    AWRAD("awrad"),
    HADITH("hadith"),
    JOURNEY("journey"),
    ;

    /** `fatwabot://prayer` */
    val uri: Uri get() = Uri.parse("$SCHEME://$host")

    companion object {
        const val SCHEME = "fatwabot"

        /** Parses an incoming URI, ignoring anything that isn't ours. */
        fun from(uri: Uri?): DeepLink? = parse(uri?.toString())

        /**
         * The actual parsing, kept off `android.net.Uri` so it is unit-testable
         * on the plain JVM — `Uri.parse` is a stubbed no-op outside an
         * instrumented/Robolectric run, which would leave the fragile half of
         * the widget→app hand-off untested.
         */
        fun parse(raw: String?): DeepLink? {
            val text = raw?.trim() ?: return null
            val separator = text.indexOf("://")
            if (separator <= 0) return null
            if (!text.substring(0, separator).equals(SCHEME, ignoreCase = true)) return null
            // Accept both `fatwabot://dua` (authority) and `fatwabot:///dua`
            // (empty authority, path only) so a malformed link still lands.
            val candidate = text.substring(separator + 3)
                .substringBefore('?')
                .substringBefore('#')
                .trim('/')
                .lowercase()
            return entries.firstOrNull { it.host == candidate }
        }
    }
}

File: ios/Packages/FatwaBotKit/Sources/CoreKit/DeepLink.swift
import Foundation

/// Canonical in-app destinations reachable from outside the app — today from
/// widgets and the Live Activity, tomorrow from push payloads or a website.
///
/// This lives in CoreKit so the app target, the widget extension and the Live
/// Activity extension all build their URLs from the SAME definition. A widget
/// that hardcodes `"fatwabot://prayer"` while the app parses `"prayer-times"`
/// fails silently — the tap just opens Home and nobody notices for weeks — so
/// the scheme, host and parsing are defined exactly once, here.
///
/// The scheme must stay in sync with `CFBundleURLSchemes` in ios/App/project.yml.
public enum DeepLink: String, CaseIterable, Sendable {
    case home
    case prayer
    case qibla
    case tasbeeh
    case azkar
    case dua
    case awrad
    case hadith
    case journey

    public static let scheme = "fatwabot"

    /// The `userInfo` key a local notification carries its route under.
    ///
    /// Lives here beside the routes themselves, not in the app target: feature
    /// modules schedule their own notifications and must be able to attach a
    /// route without depending on the app. A notification that omits it reaches
    /// the tap handler with nothing to route on and silently does nothing —
    /// which is exactly how prayer notification taps were inert.
    public static let notificationUserInfoKey = "deepLink"

    /// `fatwabot://prayer`
    public var url: URL {
        // Safe: every case is a lowercase ASCII host, so this always parses.
        URL(string: "\(Self.scheme)://\(rawValue)")!
    }

    /// Parses an incoming URL, ignoring anything that isn't ours.
    public init?(url: URL) {
        guard url.scheme?.lowercased() == Self.scheme else { return nil }
        // Accept both `fatwabot://prayer` (host) and `fatwabot:///prayer`
        // (empty host, path only) so a malformed link still lands correctly.
        let candidate = url.host?.lowercased()
            ?? url.path.trimmingCharacters(in: CharacterSet(charactersIn: "/")).lowercased()
        guard let link = DeepLink(rawValue: candidate) else { return nil }
        self = link
    }
}
